Automattic/newspack-blocks
# Newspack video management proposal & Video Playlist block

### All Submissions:






### Changes proposed in this Pull Request:



This PR is an interactive proposal for video management under Newspack. YouTube is the most common platform our publishers use, and Gutenberg has an excellent YouTube block. This PR adds a Video Playlist block. Together, these two blocks can be a complete video management and display solution.

- For single videos, embed a YouTube block.
- For featured video areas (Latest videos, a video series, etc.) use the Video Playlist block.
- The Video Playlist block can either automatically pull the latest videos from among the YouTube videos embedded in posts or can build a custom playlist of manually selected videos.
- With a block-based approach, the need for centralized video management is eliminated.
- If you want to have a video or video playlist "featured image", select the Hidden featured image style and put a video at the top of the post. It will look just like a video featured image.

### How to test the changes in this Pull Request:

We can probably hold off on code review until we've figured out the right feature set and other implementation details and are thinking about merging this.

1. Check out and build this branch.
2. Add YouTube embeds to a variety of published posts:
<img width="679" alt="Screen Shot 2020-02-07 at 10 59 36 AM" src="https://user-images.githubusercontent.com/7317227/74058488-04f2c900-499b-11ea-88d6-a85358c8e8ce.png">
3. On a new post, insert the Video Playlist block. It should auto-populate with the videos you've inserted around your site. Try the limit setting also.
<img width="602" alt="Screen Shot 2020-02-07 at 11 01 39 AM" src="https://user-images.githubusercontent.com/7317227/74058645-54d19000-499b-11ea-9e55-ca21f3fd63dc.png">
4. Test the Category filtering option by selecting a category. Only videos from posts in that category should be in the playlist.
<img width="604" alt="Screen Shot 2020-02-07 at 11 02 09 AM" src="https://user-images.githubusercontent.com/7317227/74058673-6155e880-499b-11ea-9628-46c1ffa9524f.png">
5. Test the manual mode.
<img width="678" alt="Screen Shot 2020-02-07 at 11 06 37 AM" src="https://user-images.githubusercontent.com/7317227/74058699-6c107d80-499b-11ea-9591-f10f7f839e98.png">
6. Test the Autoplay toggle. It should autoplay the video on the frontend.

### Discussion/Questions

1. On the frontend, the video playlist is sort-of hidden and you need to click the hamburger menu in the upper-right corner to display it. I don't think most people will notice that this is a playlist because of that. There is not a way to force the embed to start with that panel open as far as I can tell. It might need a custom player. Additionally, the title is always 'Untitled Playlist', which is not something that appears change-able.
<img width="721" alt="Screen Shot 2020-02-07 at 11 19 19 AM" src="https://user-images.githubusercontent.com/7317227/74058881-cf021480-499b-11ea-83b6-df9956e28c08.png">
<img width="716" alt="Screen Shot 2020-02-07 at 11 19 26 AM" src="https://user-images.githubusercontent.com/7317227/74058889-d2959b80-499b-11ea-9195-2a933c8ab2b9.png">

There are a few different approaches we could take if it turns out we need to build a custom player, but the current way the embeds are done in this PR is the simplest way, so I'm starting with it to get the discussion going.

2. This block solely supports YouTube videos. Should it be renamed to something like "YouTube Playlist"? If we support other providers in the future, will they be part of this block? Will they have their own blocks (e.g. "JWPlayer Playlist")?

3. Is that TokenField the way to go for manual YouTube playlist building? Is there a better UI/UX for that?

### Other information:




cc @deke207
